core/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -789,6 +790,27 @@ private[chisel3] object Builder extends LazyLogging {
 
   def currentWhen: Option[WhenContext] = dynamicContext.whenStack.headOption
 
+  def blockStack: List[Block] = dynamicContext.blockStack
+  def blockStack_=(s: List[Block]): Unit = {
+    dynamicContext.blockStack = s
+  }
+
+  def blockDepth: Int = dynamicContext.blockStack.length
+  def pushBlock(b: Block): Unit = {
+    dynamicContext.blockStack = b :: dynamicContext.blockStack
+  }
+
+  def popBlock(): Block = {
+    val lastBlock = dynamicContext.blockStack.head
+    dynamicContext.blockStack = dynamicContext.blockStack.tail
+    lastBlock
+  }
+
+  def currentBlock: Option[Block] = dynamicContextVar.value match {
+    case Some(dc) => dc.blockStack.headOption
+    case _        => None
+  }
+
```

Comment:
Yes, we definitely should drop the other stacks -- actually, this change is /nearly/ enough to drop the whenStack entirely, but I restored it since it's needed for now to implement `cond`.  Presently Blocks point to their owner commands, which is at different level/layer vs Chisel Data.  But it'd be great to be able to track this with one stack, yes!

Reply:
Incidentally, my original idea for how to implement `when.cond` was that when you call it, it drops a Bool wire connected to false just before the outermost when (or at the very top of the module but that'd be uglier), then connects it to true in the current when scope. But we couldn't do this because we didn't have the ability to insert at arbitrary points. Your work here would actually make that implementation possible.